src/main/scala/io/qbeast/spark/delta/writer/SparkDeltaDataWriter.scala

```diff
@@ -0,0 +1,182 @@
+
+import scala.collection.parallel.ForkJoinTaskSupport
+import scala.collection.parallel.immutable.ParVector
+
+/**
+ * Spark implementation of the DataWriter interface.
+ */
+object SparkDeltaDataWriter extends DataWriter[DataFrame, StructType, FileAction] {
+
+  override def write(
+      tableID: QTableID,
+      schema: StructType,
+      qbeastData: DataFrame,
+      tableChanges: TableChanges): IISeq[FileAction] = {
+
+    val sparkSession = qbeastData.sparkSession
+
+    val (factory: OutputWriterFactory, serConf: SerializableConfiguration) = {
+      val format = new ParquetFileFormat()
+      val job = Job.getInstance()
+      (
+        format.prepareWrite(sparkSession, job, Map.empty, schema),
+        new SerializableConfiguration(job.getConfiguration))
+    }
+
+    val qbeastColumns = QbeastColumns(qbeastData)
+
+    val blockWriter =
+      BlockWriter(
+        dataPath = tableID.id,
+        schema = schema,
+        schemaIndex = qbeastData.schema,
+        factory = factory,
+        serConf = serConf,
+        qbeastColumns = qbeastColumns,
+        tableChanges = tableChanges)
+    qbeastData
+      .repartition(col(cubeColumnName))
+      .queryExecution
+      .executedPlan
+      .execute
+      .mapPartitions(blockWriter.writeRow)
+      .collect()
+      .toIndexedSeq
+
+  }
+
+  /**
+   * Compact the files.
+   * Method based on Delta Lake OptimizeTableCommand
+   * Experimental: the implementation may change by using the CubeID as partition key
+   * and delegating the further execution to the underlying Format
+   *
+   * @param tableID
+   * @param schema
+   * @param data
+   * @param tableChanges
+   * @return
+   */
+  override def compact(
+      tableID: QTableID,
+      schema: StructType,
+      indexStatus: IndexStatus,
+      tableChanges: TableChanges): IISeq[FileAction] = {
+
+    val sparkSession = SparkSession.active
+
+    // Check what cubes are suitable for compaction
+    val cubesToCompact = indexStatus.cubesStatuses
+      .mapValues(_.files.filter(_.size >= MIN_FILE_SIZE_COMPACTION))
+      .filter(_._2.nonEmpty)
+
+    // Group the files into units of data
+    // that lays between [MIN_FILE_SIZE_COMPACTION, MAX_FILE_SIZE_COMPACTION]
+    // Based on groupFilesIntoBins from Delta Lake OptimizeTableCommand
+    val cubesToCompactGrouped = {
+      cubesToCompact.flatMap { case (cube, blocks) =>
+        val groups = Seq.newBuilder[Seq[QbeastBlock]]
+        val g = Seq.newBuilder[QbeastBlock]
+        var count = 0L
+
+        blocks.foreach(b => {
+          if (b.size + count > MAX_FILE_SIZE_COMPACTION) {
+            // If we reach the MAX_FILE_SIZE_COMPACTION limit
+            // we output a group of files for that cube
+            groups += g.result()
+            // Clear the current group
+            g.clear()
+            // Add the block to the current group
+            g += b
+            count = b.size
+          } else {
+            g += b
+            count += b.size
```

Comment:
Maybe this part can be generalized outside the If? It seems to be done regardless of the condition :)

Reply:
Yes, you are right, just need to set the count to 0 in the condition

Reply:
Done!